oss_lib/config.py

```diff
@@ -0,0 +1,235 @@
+    )
+    parser.add_argument(
+        "--log-config-file",
+        dest="log_config_path",
+        metavar="PATH",
+        help="Path to a logging configuration file, overrides the "
+             "{0}_LOG_CONF environment variable.".format(service_prefix),
+    )
+    parser.add_argument(
+        "--debug",
+        action="store_true",
+        default=False,
+        help="Set the logging level to DEBUG, overrides {0}_DEBUG "
+             "environment variable. (default: false).".format(service_prefix),
+    )
+    return parser
+
+
+def get_env_variables(service_prefix):
+    debug = os.environ.get(service_prefix + "_DEBUG") == "true"
+    variables = {
+        "config_path": os.environ.get(service_prefix + "_CONF"),
+        "log_config_path": os.environ.get(service_prefix + "_LOG_CONF"),
+        "debug": debug,
+    }
+    return variables
+
+
+def setup_config(config_path=None, default_config_path=None, defaults=None,
+                 log_config_path=None, debug=False,
+                 validation_schema=None):
+    global CONF, _CONF
+
+    setup_logging(log_config_path, debug=debug)
+
+    if defaults:
+        config = copy.deepcopy(defaults)
+    else:
+        config = {}
+
+    loaded_config = load_config_file(config_path, default_config_path)
+    if loaded_config:
+        merge_dicts(config, loaded_config)
+
+    if debug:
+        log_config(config, config_path, default_config_path)
+
+    validate_config(config, validation_schema)
+    _CONF = config
+    return CONF
+
+
+def setup_logging(config_path, debug=False):
+    if not config_path:
+        logging.basicConfig(format=LOG_FORMAT, datefmt=LOG_DATE_FORMAT)
+    else:
+        LOG.info("Logging configuration file is specified: %s",
+                 config_path)
+        logging.config.fileConfig(config_path)
+    if debug:
+        LOG.info("Setting up the DEBUG logging level.")
+        root_logger = logging.getLogger()
+        root_logger.setLevel(logging.DEBUG)
+
+
+def load_config_file(config_path, default_config_path=None):
+    config = None
+    if config_path:
+        LOG.info("Configuration file was specified: %s", config_path)
+        config = yaml_load(config_path)
+    elif default_config_path and os.path.exists(default_config_path):
+        LOG.info("Trying to read default configuration file: %s",
+                 default_config_path)
+        config = yaml_load(default_config_path)
+    return config
+
+
+def yaml_load(path):
+    try:
+        with open(path) as f:
+            return yaml.safe_load(f)
+    except IOError as exc:
+        LOG.error("Could not read content from %s: %s", path, exc)
+        raise
+    except yaml.YAMLError as exc:
+        LOG.error("Could not parse YAML-content from %s: %s", path, exc)
+        raise
+
+
+def merge_dicts(source, other):
+    for key, other_value in six.iteritems(other):
+        try:
+            value = source[key]
+        except KeyError:
```

Comment:
We can not just replace an old value by a new one, we need `value` after it if it is a `dict` to properly merge inner `dict`s and do not just replace one by one. Otherwise, it will ruin the whole idea of default values.